File: dotnet/samples/SimpleOrm.Sample/Migrations/Table/User/V0002_AddDisplayName.cs
namespace SimpleOrm.Sample.Migrations.Table.User;

/// <summary>
/// A real change migration: literal column spec (frozen forever), with the
/// per-action Post hook backfilling existing rows — data work rides the same
/// version atomicity as the DDL.
/// </summary>
public sealed class V0002_AddDisplayName : TableMigration<Models.User>
{
    public override void Action(TableActions actions) => actions
        .AddColumn("display_name", "TEXT")
        .Post("update users set display_name = name");

    // No hand-written Down (owner decision, ADR-0016): rollbacks derive from the
    // versioned schema snapshots (diff V0002 vs V0001) once the generator exists;
    // until then, migrate down refuses honestly with MIG-020.
}

File: php/samples/SimpleOrm.Sample/src/Migrations/Table/User/V0002_AddDisplayName.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Sample\Migrations\Table\User;

use SimpleOrm\Migrations\TableActions;
use SimpleOrm\Migrations\TableMigration;
use SimpleOrm\Sample\Models\Tables\User;

/**
 * A real change migration: literal column spec (frozen forever), with the
 * per-action post hook backfilling existing rows. Data work rides the same
 * version atomicity as the DDL. No hand-written down (ADR-0016/0018): the
 * rollback derives from the versioned schema snapshots.
 */
final class V0002_AddDisplayName extends TableMigration
{
    public function entityClass(): string
    {
        return User::class;
    }

    public function action(TableActions $actions): void
    {
        $actions
            ->addColumn('display_name', 'TEXT')
            ->post('update users set display_name = name');
    }
}
